librad/src/id/entity.rs

```diff
@@ -0,0 +1,614 @@
+    pub async fn sign(
+        &mut self,
+        key: &Key,
+        by: &Signatory,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        let public_key = key.public();
+        if self.signatures().contains_key(&public_key) {
+            return Err(Error::SignatureAlreadyPresent(public_key.to_owned()));
+        }
+        self.check_key(&public_key, by, resolver).await?;
+        let signature = EntitySignature {
+            by: by.to_owned(),
+            sig: self.compute_signature(key)?,
+        };
+        self.signatures.insert(public_key, signature);
+        Ok(())
+    }
+
+    /// Check that an entity signature is valid
+    pub async fn check_signature(
+        &self,
+        key: &PublicKey,
+        by: &Signatory,
+        signature: &Signature,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        self.check_key(key, by, resolver).await?;
+        if signature.verify(&self.canonical_data()?, key) {
+            Ok(())
+        } else {
+            Err(Error::SignatureVerificationFailed)
+        }
+    }
+
+    /// Check the validity of this entty (only this revision is checked)
+    ///
+    /// This checks that:
+    /// - every owned key and certifier has a corresponding signature
+    /// - only owned keys and certifiers have signed the entity
+    pub async fn check_validity(&self, resolver: &impl Resolver<User>) -> Result<(), Error> {
+        let mut keys = HashSet::<PublicKey>::from_iter(self.keys().iter().cloned());
+        let mut users = HashSet::<RadicleUri>::from_iter(self.certifiers().iter().cloned());
+
+        for (k, s) in self.signatures() {
+            self.check_signature(k, &s.by, &s.sig, resolver).await?;
+            match &s.by {
+                Signatory::OwnedKey => {
+                    keys.remove(k);
+                },
+                Signatory::User(user) => {
+                    users.remove(&user);
+                },
+            }
+        }
+        if keys.is_empty() && users.is_empty() {
+            Ok(())
+        } else {
+            Err(Error::SignatureMissing)
+        }
+    }
+
+    /// Convenience method to check if an entity is valid
+    pub async fn is_valid(&self, resolver: &impl Resolver<User>) -> bool {
+        self.check_validity(resolver).await.is_ok()
+    }
+
+    /// Given an entity and its previous revision check that the update is
+    /// valid:
+    ///
+    /// - the revision has been incremented
+    /// - the parent hash is correct
+    /// - the TUF quorum rules have been observed
+    ///
+    /// FIXME: only allow exact `+1`increments so that the revision history has
+    /// no holes
+    /// FIXME: probably we should merge owned keys and certifiers when checking
+    /// the quorum rules (now we are handling them separately)
+    pub fn check_update(&self, previous: &Self) -> Result<(), UpdateVerificationError> {
+        if self.revision() <= previous.revision() {
```

Comment:
You mention this in the comment, but I think it's pretty important that the updates are monotonic.

Reply:
Why do you think it is important? I understand Massi's point more as an API ergonomics concern.

The reason TUF implies that there be no holes in the sequence is so you can't specify an arbitrarily large one. Otoh, this may make stalemate situations more complicated to resolve -- our system can by (ab-)used for voting schemes, which is not a consideration for TUF.

We can start with adhering strictly to TUF (next revision must be exactly previous + 1), and loosen it to a (small) window later.

Reply:
Yeah a couple things come to mind, though maybe we already have solutions for them:
1) Do we have a way to tell if we're missing an update, between eg. Rev 31 and Rev 12319?
2) Are there overflow attacks we should look out for?
3) Given two updates to the same parent state, can we tell that they are conflicting (and not sequences), even though they may have different revisions?

I think it makes things easier to think about at the very least, if you know ahead of time what the revision number of the next update is going to be, given that updates can be made concurrently.

Reply:
> Do we have a way to tell if we're missing an update, between eg. Rev 31 and Rev 12319?

I think in this case you should be able to tell since the hash doesn't match, ala Merkle trees and all that. But someone can correct me if I'm wrong here.

Reply:
Yes it's all hash-linked -- which is **not** the case in TUF. The only reason really we have this number is that it seemed unclear how one would reify the order in a patch-based system, but I know think that it would be through native means.

Keep in mind that the logic here will verify only _one_ peer's view of a given identity document. In order to reach a "consensus" (e.g. when talking about what the `master` head is), the views of all maintainers have to agree. If they don't the revision number will make things more complicated: say there's a 50-50 split between two camps of maintainers at proposed rev 42, but the tiebreaker doesn't agree with neither of them, they can't just propose rev 43, because that would violate the gap-less condition. This is quite straightforward if that condition was lifted, or there would be no revision at all.

Reply:
Gotcha, yeah I didn't catch that it was hash linked.

Reply:
Having revision numbers with a strict `+1` rule will make key revocation checks way simpler.
I would discuss this in the "key revocation handling" PR I will post as soon as this is merged (see #87).)

Reply:
Please consider removing the numeric revision altogether, for the reasons outlined above. Since we said we'll need a cache to make lookups O(1), we could just as well key by hash.